app/services/container_dashboard_service.rb

```diff
@@ -0,0 +1,206 @@
+          :count     => ManageIQ::Providers::Kubernetes::ContainerManager.count,
+        }
+      ]
+    }
+  end
+
+  def node_utilization
+    cpu_total = 0
+    mem_total = 0
+
+    if @ems.present?
+      providers = [@ems]
+    else
+      providers = (ManageIQ::Providers::Openshift::ContainerManager.all <<
+                     ManageIQ::Providers::Kubernetes::ContainerManager.all).flatten
+    end
+
+    providers.each do |prov|
+      cpu_total += prov.aggregate_logical_cpus # get this from metrics as well?
+      mem_total += prov.aggregate_memory       # get this from metrics as well?
+    end
+
+    mem_total = memory_to_human_size(mem_total).to_i # Always in GB?
+
+    {
+      :cpu => {
+        :total => cpu_total,
+        :used  => cpu_total - 3, # Dummy data
+        :xData => (["dates"] << (1..30).to_a).flatten, # Dummy data
+        :yData => (["used"] << (1..30).collect { |_x| rand(cpu_total) }).flatten # Dummy data
+      },
+      :memory => {
+        :total => mem_total,
+        :used  => mem_total - 7, # Dummy data
+        :xData => (["dates"] << (1..30).to_a).flatten, # Dummy data
+        :yData => (["used"] << (1..30).collect { |_x| rand(mem_total) }).flatten # Dummy data
+      }
+    }
+  end
+
+  def memory_to_human_size(mem)
+    ActionView::Base.new.number_to_human_size(mem * 1.megabyte, :precision => 0).split[0]
+  end
+
+  def network_utilization
+    {
+      :current => {
+        :xData => (["dates"] << (0..15).to_a).flatten, # Dummy data - should have dates\timestamps as elements
+        :yData => (["data"] << (0..15).collect { |_x| rand(1200)}).flatten # Dummy data
+      },
+      :monthly => {
+        :xData => (["dates"] << (0..30).to_a).flatten, # Dummy data
+        :yData => (["data"] << (0..30).collect { |_x| rand(1200)}).flatten # Dummy data
+      }
+    }
+  end
+
+  def entity_trends
+    {
+      :pods => {
+        :xData => (["dates"] << (0..30).to_a).flatten, # Dummy data - should have dates\timestamps as elements
+        :yData => (["pods"] << (0..30).collect { |x| x + rand(x)}).flatten # Dummy data
+      },
+      :images => {
+        :xData => (["dates"] << (0..30).to_a).flatten, # Dummy data - should have dates\timestamps as elements
+        :yData => (["images"] << (0..30).collect { |x| x + rand(x)}).flatten # Dummy data
+      }
+    }
+  end
+
+  def heatmaps
+    if @ems.present?
+      node_ids = @ems.container_nodes.pluck(:id)
+    else
+      node_ids = ContainerNode.all.pluck(:id)
+    end
+
+    # (What about nodes that have no metrics?)
+    # Get last Metric for each node
+    node_metrics = Metric.find_by_sql(
+     "SELECT *
+      FROM   (SELECT m.*, n.*, ROW_NUMBER() OVER(PARTITION BY resource_id ORDER BY timestamp DESC) AS rk
+              FROM   metrics AS m
+              JOIN   container_nodes AS n ON resource_id = n.id) t
+      WHERE  rk = 1").select {|m| node_ids.include? m.resource_id} # Add this part to sql query?
```

Comment:
> Since you have timestamp, can't you just use MAX(timestamp) to get the latest node.

I tried to evade having a query for each node. 
```Ruby
 ContainerNode.each { |n| n.metrics.where....}
```

I was afraid of the overhead. Is this solution unnecessary?